applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -919,6 +929,116 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::ComputePotentia
+
+    // get node IDs of upwind element
+    int upwind_element_one_node_id = 0;
+    int upwind_element_two_node_id = 0;
+
+    // vector between points 0 and 1 has biggest streamwise component
+    if(first_edge_normal_fs_comp > second_edge_normal_fs_comp && first_edge_normal_fs_comp > third_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 0;
+        upwind_element_two_node_id = 1;
+    }
+    // vector between points 1 and 2 has biggest streamwise component
+    else if(second_edge_normal_fs_comp > first_edge_normal_fs_comp && second_edge_normal_fs_comp > third_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 1;
+        upwind_element_two_node_id = 2;
+    }
+    // vector between points 0 and 2 has biggest streamwise component
+    else if(third_edge_normal_fs_comp > second_edge_normal_fs_comp && third_edge_normal_fs_comp > first_edge_normal_fs_comp)
+    {
+        upwind_element_one_node_id = 0;
+        upwind_element_two_node_id = 2;
+    }
+    else
+    {
+         KRATOS_WARNING("Transonic perturbation element: FindUpWindElement") <<
+        "no maximum streamwise component found" << std::endl;
+    }
+
+    // find elements attached to each node
+    const GlobalPointersVector<Element>& rNodeOneElementCandidates = r_geom[upwind_element_one_node_id].GetValue(NEIGHBOUR_ELEMENTS);
+    const GlobalPointersVector<Element>& rNodeTwoElementCandidates = r_geom[upwind_element_two_node_id].GetValue(NEIGHBOUR_ELEMENTS);
+
+    bool loop_stop = false;
+
+    // find element which shares both nodes
+    for (SizeType i = 0; i < rNodeOneElementCandidates.size() && !loop_stop; i++)
+    {
+        for (SizeType j = 0; j < rNodeTwoElementCandidates.size() && !loop_stop; j++)
+        {
+            if(rNodeOneElementCandidates(i)->Id() == rNodeTwoElementCandidates(j)->Id() && rNodeOneElementCandidates(i)->Id() != r_this.Id())
+            {
+                // assign upwind element
+                mpUpwindElement = rNodeOneElementCandidates(i);
+
+                // get current element and upwind element center points
+                const GeometryType& r_upwind_element_geometry = rNodeOneElementCandidates(i)->GetGeometry();
+                const Point upwind_element_center = r_upwind_element_geometry.Center();
+
+                // make vector pointing from current element to upwind element
+                vector<double> vector_to_upwind_element (3, 0.0);
+                vector_to_upwind_element[0] = upwind_element_center[0] - current_element_center[0];
+                vector_to_upwind_element[1] = upwind_element_center[1] - current_element_center[1];
+                vector_to_upwind_element[2] = upwind_element_center[2] - current_element_center[2];
+
+                this->SetValue(VECTOR_TO_UPWIND_ELEMENT, vector_to_upwind_element);
+
+                loop_stop = true;
+                break;
+            }
+        }
+    }
+}
```

Comment:
Do you think this is a good idea? Let's discuss it later
```suggestion
    // If no upwind element is found, the element is INLET and the
    // upwind element pointer points to itself.
    if (mpUpwindElement.get() == nullptr) {
        mpUpwindElement = this;
        this->Set(INLET);
    }
}
```